lesson06.md

```diff
@@ -0,0 +1,391 @@
+
+In such situations (and many more), JavaScript will stop its execution and show an error:
+
+- trying to use a variable that does not exist
+- trying to define two variables with the same name
+- omitting parenthesis in an if...else statement
+- invalid operations like `2 = 2`
+
+---
+
+### Debugging: intro
+
+```js
+console.log(fullName);
+
+let firstName = "Carlo";
+let firstName = "Harald";
+
+let isRaining = true;
+if (isRaining === true {
+    console.log("take an umbrella");
+}
+
+2 = 2;
+
+```
+
+---
+
+### Debugging: intro
+
+Some errors are more subtle. 
+
+Code is valid, it looks correct, but it behaves in unexpected ways.
+
+```js
+let temperature = 37;
+
+if (temperature > 20) {
+  console.log("it's warm");
+} else if (temperature > 30) {
+  console.log("too hot!!!");
+} else if (temperature > 5) {
+  console.log("perfect.");
+} else {
+  console.log("it's a bit chilly.")
+}
+```
+
+---
+
+### Debugging: intro
+
+More often than not, bugs are errors in the logic of our programs.
+
+- hard to even realise there is a bug in the first place
+    - JavaScript can't warn us
+    - we don't notice them until something goes wrong or unexpectedly
+- hard to find where the bug is in our code
+    - a bug can be caused by a completely different line in our code
+- hard to fix, even after realising we have a bug and finding where it is
+
+---
+
+### Debugging: intro
+
+In an ideal world we would be able to pause the execution of our program at any point in time and check everything that's going on.
+
+And surprisingly, we live in that world!
+<!-- .element: class="fragment" -->
+
+---
+
+### Debugging: the debugger keyword 
+
+We can use the `debugger` keyword to tell JavaScript where to pause its execution.
+
+```js
+let a = 0;
+a++;
+debugger;
+let b = a;
+```
+
+---
+
+### Debugging: the debugger 
+
+If our code contain a `debugger` statement and we open the browser console, we'll see this thing on top:
+
+![](images/debugger.png) <!-- .element: style="display: block; margin: auto;" -->
+
+With this, we can follow JavaScript's execution.
+
+---
+
+### Debugging: the debugger 
+
+If our code contain a `debugger` statement and we open the browser console, we'll see this thing on top:
+
+![](images/debugger.png) <!-- .element: style="display: block; margin: auto;" -->
+
+With this, we can follow JavaScript's execution.
+
+---
+
+### Debugging: the debugger
+
+![](images/debugger-full-view.png) <!-- .element: style="width: 1000px; display: block; margin: auto;" -->
+
+---
+
+### Debugging: the debugger
+
+When we pause JavaScript's execution, we can inspect the status of our program at that moment in time.
+
+We can also use the `Watch` panel on the right to keep an eye on variables or whole expressions and see how they change.
+
+---
+
+### Debugging: try it yourself 
+
+- Add a debugger statement somewhere in the code below and run it.
+- add some variables and/or expression to `watch` and see their values changing
+
```

Comment:
Maybe add something like "be sure to remove the debugger statement when you are happy with how the code works", since it's a bit more important to clean up than console.log

Reply:
good point!